flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogKeyedStateBackend.java

```diff
@@ -565,18 +568,44 @@ public <N, SV, SEV, S extends State, IS extends S> IS createInternalState(
                     StateSnapshotTransformer.StateSnapshotTransformFactory<SEV>
                             snapshotTransformFactory)
             throws Exception {
+        ChangelogState changelogState =
+                changelogStateFactory.getExistingState(
+                        stateDesc.getName(), BackendStateType.KEY_VALUE);
+        if (changelogState == null) {
```

Comment:
I think it should also work even if the createInternalState is called firstly because it will trigger the migration normally.
In this case, the upgrade method will be called when user access.

Reply:
I have modified the `upgrade` method and make restore procedure always call `upgrade`.
So currently, `upgrade` may just create a new state or trigger a state migration and always make the metaInfo could continue to migrate.